src/napari_graph/directed_graph.py

```diff
@@ -0,0 +1,418 @@
+
+"""
+Directed edge constants for accessing the directed graph buffer data.
+Each edge occupies _DI_EDGE_SIZE spaces on the graph buffer.
+_LL_DI_EDGE_POS indicates the displacement between the edge initial index and
+the **target** edge directed linked list position.
+
+Example of a directed graph edge buffer:
+[
+    source_node_buffer_id_0,
+    target_node_buffer_id_0,
+    source_edge_linked_list_0,
+    target_edge_linked_List_0,
+    source_node_buffer_id_1,
+    target_node_buffer_id_1,
+    source_edge_linked_list_1,
+    target_edge_linked_List_1,
+    ...
+]
+"""
+_DI_EDGE_SIZE = _UN_EDGE_SIZE + 1
+_LL_DI_EDGE_POS = 2
+
+
+@njit
+def _add_directed_edge(
+    buffer: np.ndarray,
+    node2src_edge: np.ndarray,
+    node2tgt_edge: np.ndarray,
+    empty_idx: int,
+    src_node: int,
+    tgt_node: int,
+) -> int:
+    """Add a single directed edge to `buffer`.
+
+    This updates the `buffer`'s source and target linked list
+    and the nodes to edges mappings.
+
+    NOTE: see `_add_undirected_edge` docs for comment about cache misses.
+
+     Returns
+    -------
+    int
+        Next empty edge index position.
+    """
+    if empty_idx == _EDGE_EMPTY_PTR:
+        raise ValueError("Edge buffer is full.")
+
+    elif empty_idx < 0:
+        raise ValueError("Invalid empty index.")
+
+    next_src_edge = node2src_edge[src_node]
+    next_tgt_edge = node2tgt_edge[tgt_node]
+    node2src_edge[src_node] = empty_idx
+    node2tgt_edge[tgt_node] = empty_idx
+
+    buffer_index = empty_idx * _DI_EDGE_SIZE
+    next_empty = buffer[buffer_index + _LL_DI_EDGE_POS]
+
+    buffer[buffer_index] = src_node
+    buffer[buffer_index + 1] = tgt_node
+    buffer[buffer_index + _LL_DI_EDGE_POS] = next_src_edge
+    buffer[buffer_index + _LL_DI_EDGE_POS + 1] = next_tgt_edge
+
+    return next_empty
+
+
+@njit
+def _add_directed_edges(
+    buffer: np.ndarray,
+    edges: np.ndarray,
+    empty_idx: int,
+    n_edges: int,
+    node2src_edge: np.ndarray,
+    node2tgt_edge: np.ndarray,
+) -> Tuple[int, int]:
+    """Add an array of edges into the `buffer`.
+
+    Directed edges contains two linked lists, outgoing (source) and incoming
+    (target) edges.
+    """
+    size = edges.shape[0]
+    for i in range(size):
+
+        empty_idx = _add_directed_edge(
+            buffer,
+            node2src_edge,
+            node2tgt_edge,
+            empty_idx,
+            edges[i, 0],
+            edges[i, 1],
+        )
+        n_edges += 1
+
+    return empty_idx, n_edges
+
+
+@njit(inline='always')
+def _remove_target_edge(
+    src_node: int,
+    tgt_node: int,
+    edges_buffer: np.ndarray,
+    node2tgt_edges: np.ndarray,
+) -> None:
+    """Remove edge from target edges linked list.
+
+    It doesn't clean the buffer, because it'll be used later.
+    """
+
+    idx = node2tgt_edges[tgt_node]  # different indexing from source edge
+    prev_buffer_idx = _EDGE_EMPTY_PTR
+
+    while idx != _EDGE_EMPTY_PTR:
```

Comment:
As noted elsewhere, can this be converted to a for loop to ensure that it'll terminate? One example could be `for i in range(edges_buffer.shape[0]):`, since it's guaranteed that you will have looked at every position by then *or* you'll be in an infinite loop. ;)